pil-analyzer/tests/condenser.rs

```diff
@@ -471,3 +484,158 @@ fn intermediate_arr_wrong_length() {
 "#;
     analyze_string::<GoldilocksField>(input);
 }
+
+#[test]
+fn closure() {
+    let input = r#"
+    namespace std::prover;
+        let eval = 8;
+    namespace N(16);
+        col witness x;
+        let y = |a, b, c| Query::Hint(a + c());
+        {
+            let r = 9;
+            set_hint(x, |i| y(1, i, || 9 + r));
+        };
+"#;
+    let analyzed = analyze_string::<GoldilocksField>(input);
+    let expected = r#"namespace std::prover;
+    let eval = 8;
+namespace N(16);
+    col witness x;
+    std::prelude::set_hint(N::x, {
+        let r = 9;
+        (|i| N::y(1, i, (|| 9 + r)))
+    });
+    let y: fe, int, (-> fe) -> std::prelude::Query = (|a, b, c| std::prelude::Query::Hint(a + c()));
+"#;
+    assert_eq!(analyzed.to_string(), expected);
+}
+
+#[test]
+fn closure_complex() {
+    let input = r#"
+    namespace std::prover;
+        let eval = 8;
+    namespace std::convert;
+        let fe = 9;
+    namespace N(16);
+        col witness x;
+        let y = |a, b, c| Query::Hint(a + c());
+        {
+            let r = 9;
+            let k: int[] = [-2];
+            let q = "" != "";
+            let b = (|s| || "" == s)("");
+            set_hint(x, |i| {
+                y(1, i, || if b() && q { 9 + r } else { std::convert::fe(k[0]) })
+            });
+        };
+"#;
+    let analyzed = analyze_string::<GoldilocksField>(input);
+    let expected = r#"namespace std::prover;
+    let eval = 8;
+namespace std::convert;
+    let fe = 9;
+namespace N(16);
+    col witness x;
+    std::prelude::set_hint(N::x, {
+        let r = 9;
+        let k = [-2];
+        let q = std::prelude::false;
+        let b = {
+            let s = "";
+            (|| "" == s)
+        };
+        (|i| { N::y(1, i, (|| if b() && q { 9 + r } else { std::convert::fe::<int>(k[0]) })) })
+    });
+    let y: fe, int, (-> fe) -> std::prelude::Query = (|a, b, c| std::prelude::Query::Hint(a + c()));
+"#;
+    assert_eq!(analyzed.to_string(), expected);
+}
+
+#[test]
+fn prover_functions() {
+    let input = "
+    namespace std::convert;
+        let fe = 8;
+    namespace std::prover;
+        let provide_value = 9;
+    namespace N(16);
+        let x;
+        let y;
+        x = y;
+        query |i| {
+            std::prover::provide_value(x, i, std::convert::fe(i % 2));
+            std::prover::provide_value(y, i, std::convert::fe(i % 2));
+        };
+    ";
+    let analyzed = analyze_string::<GoldilocksField>(input);
+    let expected = r#"namespace std::convert;
+    let fe = 8;
+namespace std::prover;
+    let provide_value = 9;
+namespace N(16);
+    col witness x;
+    col witness y;
+    N::x = N::y;
+    (query |i| {
+        std::prover::provide_value(N::x, i, std::convert::fe::<int>(i % 2));
+        std::prover::provide_value(N::y, i, std::convert::fe::<int>(i % 2));
+
+    });
```

Comment:
I appears to me that `set_hint` and `provide_value` works differently in that `set_hint` "captures" the context and moves it into the lambda whereas `provide_value` directly uses context without moving it into the lambda. Is it a deliberate design difference? Are we removing `set_hint` eventually?

Reply:
Yes, the idea is to replace `set_hint` by query functions that can provide hints to multiple columns at the same time. This is useful since often, multiple columns depend on the same intermediate value being computed. Without that, we would have to re-compute the intermediate value multiple times.

Reply:
Makes sense thanks :)